tempesta_fw/pool.c

```diff
@@ -35,33 +36,65 @@
 TfwPool *
 __tfw_pool_new(size_t n)
 {
-	unsigned int order = (n + sizeof(TfwPool)) >> PAGE_SHIFT;
+	unsigned int order = get_order(n + sizeof(TfwPool) + sizeof(TfwPoolChunk));
 	TfwPool *p;
+	TfwPoolChunk *chunk;
 
 	p = (TfwPool *)__get_free_pages(GFP_ATOMIC, order);
 	if (!p)
 		return NULL;
-	p->base = (unsigned char *)p;
-	p->order = order;
-	p->off = sizeof(*p);
+
+	chunk = (TfwPoolChunk *)(p + 1);
+	chunk->base = (unsigned char *)p;
+	chunk->order = order;
+	chunk->off = sizeof(*p) + sizeof(*chunk);
+	chunk->next = NULL;
+	p->head = chunk;
 
 	return p;
 }
 EXPORT_SYMBOL(__tfw_pool_new);
 
+TfwPoolChunk *
+__tfw_find_chunk(TfwPool *p, void *ptr)
+{
+	unsigned char *tmp_ptr = ptr;
+	TfwPoolChunk *chunk;
+
+	chunk = p->head;
+	while (chunk != NULL) {
+		if (chunk->base <= tmp_ptr &&
+		    tmp_ptr < chunk->base + TFW_POOL_CHUNK_SIZE(chunk)) {
+			break;
+		}
+		chunk = chunk->next;
+	}
```

Comment:
You don't need linear scan. The chunk can be retrieved simply by ```(unsigned long)ptr & PAGE_MASK)``` since TfwPoolChunk is always at head of the page where ptr is.